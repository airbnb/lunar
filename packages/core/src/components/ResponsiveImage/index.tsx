import React from 'react';
import { mutuallyExclusiveTrueProps } from 'airbnb-prop-types';
import withStyles, { css, WithStylesProps } from '../../composers/withStyles';

export const DEFAULT_BORDER_RADIUS = 6;
const objectFitPropType = mutuallyExclusiveTrueProps('contain', 'cover');

export type Props = {
  /** An accessible label. */
  alt: string;
  /** The css border-radius value. Defaults to 6 pixels. */
  borderRadius?: number;
  /** Hide the box-shadow. */
  noShadow?: boolean;
  /** Whether to have the image maintain its aspect ratio while contained in its dimentions or not. */
  contain?: boolean;
  /** Whether to have the image cover its dimensions or not. */
  cover?: boolean;
  /**
   * The maximum width of the image (and the component). Unconstrained (css value 'none') by
   * default. If the image is wider than the value given for maxWidth, it will be scaled
   * proportionally to be at most maxWidth wide. The image width will be less than maxWidth
   * if either its native size is smaller than this value or its height exceeds maxHeight and
   * proportionally scaling to fit maxHeight produces a width less than maxWidth.
   */
  maxWidth?: string | number;
  /**
   * The maximum height of the image (and the component). Unconstrained (css value 'none') by
   * default. If the image is taller than the value given for maxHeight, it will be scaled
   * proportionally to be at most maxHeight high. The image height will be less than maxHeight
   * if either its native size is smaller than this value or its width exceeds maxWidth and
   * proportionally scaling to fit maxWidth produces a height less than maxHeight.
   */
  maxHeight?: string | number;
  /** The source of the image. Typically a URL. */
  src: string;
  /** A component to be displayed while the the image is loading. Typically a Shimmer. */
  shimmer?: React.ReactNode;
};

export type State = {
  imageLoaded: boolean;
};

/** An image that is constrained proportionally in one or both dimensions. */
export class ResponsiveImage extends React.Component<Props & WithStylesProps, State> {
  static defaultProps = {
    borderRadius: DEFAULT_BORDER_RADIUS,
    contain: false,
    cover: false,
    maxHeight: 'none',
    maxWidth: 'none',
    noShadow: false,
    shimmer: null,
  };

  static propTypes = {
    contain: objectFitPropType,
    cover: objectFitPropType,
  };

  image?: HTMLImageElement;

  state = { imageLoaded: false };

  componentDidMount() {
    this.createAsyncImage();
  }

  componentDidUpdate({ src: previousSrc }: Props) {
    const { src: currentSrc } = this.props;

    if (currentSrc !== previousSrc) {
      this.removeImage();
      this.setState({ imageLoaded: false });
      this.createAsyncImage();
    }
  }

  componentWillUnmount() {
    this.removeImage();
  }

  createAsyncImage() {
    this.image = new Image();
    this.image.addEventListener('load', () => {
      this.handleAsyncImageLoad();
    });
    this.image.addEventListener('error', () => {
      this.handleAsyncImageLoad();
    });
    this.image.src = this.props.src;
  }

  removeImage() {
    if (this.image) {
      // Prevent these callbacks from being fired if they haven't fired yet.
      delete this.image.onload;
      delete this.image.onerror;
    }
  }

  private handleAsyncImageLoad() {
    // We no longer need the Image object, so let's remove it so it can be
    // garbage collected.
    delete this.image;

    this.setState({ imageLoaded: true });
  }

  render() {
    const {
      alt,
      contain,
      cover,
      noShadow,
      maxWidth,
      maxHeight,
      src,
      borderRadius,
      shimmer,
      styles,
    } = this.props;
    const { imageLoaded } = this.state;

    if (!imageLoaded && shimmer) {
      return shimmer;
    }

    return (
      <img
        {...css(
          styles.image,
          contain && styles.image_contain,
          cover && styles.image_cover,
          noShadow && styles.image_noShadow,
          {
            borderRadius,
            maxWidth,
            maxHeight,
          },
        )}
        src={src}
        width="100%"
        height="auto"
        alt={alt}
      />
    );
  }
}

export default withStyles(({ ui }) => ({
  image: {
    display: 'block',
    boxShadow: ui.boxShadow,
  },

  image_noShadow: {
    boxShadow: 'none',
  },

  image_cover: {
    objectFit: 'cover',
  },

  image_contain: {
    objectFit: 'contain',
  },
}))(ResponsiveImage);
