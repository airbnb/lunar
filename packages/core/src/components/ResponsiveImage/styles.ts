import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ ui }) => ({
  image: {
    display: 'block',
    boxShadow: ui.boxShadow,
  },

  image_noShadow: {
    boxShadow: 'none',
  },

  image_cover: {
    objectFit: 'cover',
  },

  image_contain: {
    objectFit: 'contain',
  },
});

export default styleSheet;

export { styleSheet };
