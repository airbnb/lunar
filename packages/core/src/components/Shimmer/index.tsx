import React from 'react';
import withStyles, { WithStylesProps } from '../../composers/withStyles';

export type Props = {
  /** Display as block instead of inline. */
  block?: boolean;
  /** Height of the bar. */
  height?: number | string;
  /** Width of the bar. Pass "random" to generate a random width. */
  width?: number | string;
  /** Border radius of the bar. */
  radius?: string;
};

/** A loading indicator bar that shimmers. */
export class Shimmer extends React.Component<Props & WithStylesProps> {
  static defaultProps = {
    block: false,
    height: '1.25ex',
    radius: '0.125em',
    width: '60%',
  };

  randomWidth: string =
    this.props.width === 'random' ? `${Math.round(Math.random() * (90 - 30) + 30)}%` : '';

  render() {
    const { cx, block, height, radius, styles, width } = this.props;

    return (
      <span
        aria-busy="true"
        style={{
          borderRadius: radius,
          width: this.randomWidth || width,
          height,
        }}
        className={cx(styles.shimmer, block && styles.shimmer_block)}
      />
    );
  }
}

export default withStyles(() => ({
  shimmer: {
    animationDirection: 'alternate',
    animationDuration: '1s',
    animationFillMode: 'forwards',
    animationIterationCount: 'infinite',
    animationName: {
      name: 'shimmer',
      from: { opacity: 0.1 },
      to: { opacity: 0.3 },
    },
    animationTimingFunction: 'ease-in-out',
    backgroundColor: 'currentColor',
    display: 'inline-block',
    position: 'relative',
    verticalAlign: 'middle',
  },

  shimmer_block: {
    display: 'block',
  },
}))(Shimmer);
