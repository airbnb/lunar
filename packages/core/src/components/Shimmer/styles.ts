import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = () => ({
  shimmer: {
    animationDirection: 'alternate',
    animationDuration: '1s',
    animationFillMode: 'forwards',
    animationIterationCount: 'infinite',
    animationName: {
      name: 'shimmer',
      from: { opacity: 0.1 },
      to: { opacity: 0.3 },
    },
    animationTimingFunction: 'ease-in-out',
    backgroundColor: 'currentColor',
    display: 'inline-block',
    position: 'relative',
    verticalAlign: 'middle',
  },

  shimmer_block: {
    display: 'block',
  },
});

export default styleSheet;

export { styleSheet };
