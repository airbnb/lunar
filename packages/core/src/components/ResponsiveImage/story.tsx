import React from 'react';
import lunar from ':storybook/images/lunar-logo.png';
import Shimmer from '../Shimmer';
import ResponsiveImage from '.';

export default {
  title: 'Core/ResponsiveImage',
  parameters: {
    inspectComponents: [ResponsiveImage],
  },
};

export function anImageThatIsConstrainedProportionally() {
  return (
    <ResponsiveImage
      src={lunar}
      alt="Image"
      maxWidth={150}
      maxHeight={150}
      borderRadius={6}
      shimmer={<Shimmer block height={150} width={150} />}
    />
  );
}

anImageThatIsConstrainedProportionally.story = {
  name: 'An image that is constrained proportionally.',
};

export function withContain() {
  return (
    <ResponsiveImage
      contain
      src={lunar}
      alt="Image"
      maxWidth={150}
      maxHeight={100}
      shimmer={<Shimmer block height={100} width={150} />}
    />
  );
}

withContain.story = {
  name: 'With contain.',
};

export function withCover() {
  return (
    <ResponsiveImage
      cover
      src={lunar}
      alt="Image"
      maxWidth={150}
      maxHeight={100}
      shimmer={<Shimmer block height={100} width={150} />}
    />
  );
}

withCover.story = {
  name: 'With cover.',
};

export function withNoShadow() {
  return (
    <ResponsiveImage
      noShadow
      src={lunar}
      alt="Image"
      maxWidth={150}
      shimmer={<Shimmer block height={150} width={150} />}
    />
  );
}

withNoShadow.story = {
  name: 'With no shadow.',
};
