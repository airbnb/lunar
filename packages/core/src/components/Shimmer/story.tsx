import React from 'react';
import Shimmer from '.';

export default {
  title: 'Core/Shimmer',
  parameters: {
    inspectComponents: [Shimmer],
  },
};

export function aShimmerWithRandomWidthsDefault() {
  return (
    <>
      <Shimmer block width="random" />
      <br />
      <Shimmer block width="random" />
      <br />
      <Shimmer block width="random" />
    </>
  );
}

aShimmerWithRandomWidthsDefault.story = {
  name: 'A shimmer with random widths (default).',
  parameters: { happo: false },
};

export function anInlineShimmerWithFixedDimensions() {
  return <Shimmer width={150} height={25} />;
}

anInlineShimmerWithFixedDimensions.story = {
  name: 'An inline shimmer with fixed dimensions.',
};

export function aCircleShimmerUsingBorderRadius() {
  return <Shimmer width={50} height={50} radius="50%" />;
}

aCircleShimmerUsingBorderRadius.story = {
  name: 'A circle shimmer using border radius.',
};
