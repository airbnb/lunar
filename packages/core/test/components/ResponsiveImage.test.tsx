import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import ResponsiveImage from '../../src/components/ResponsiveImage';
import Shimmer from '../../src/components/Shimmer';

describe('<ResponsiveImage />', () => {
  const props = {
    src: 'foo',
    alt: 'bar',
    maxWidth: 200,
    maxHeight: 150,
  };

  describe('with a shimmer prop', () => {
    const shimmer = <Shimmer block />;

    it('renders the shimmer component while the image is loading', () => {
      const wrapper = shallowWithStyles(<ResponsiveImage shimmer={shimmer} {...props} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('renders the image with state={ imageLoaded: true }', () => {
      const wrapper = shallowWithStyles(<ResponsiveImage shimmer={shimmer} {...props} />);
      wrapper.setState({ imageLoaded: true });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('without a shimmer prop', () => {
    it('renders the image element while the image is loading', () => {
      const wrapper = shallowWithStyles(<ResponsiveImage {...props} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('renders the image element with state={ imageLoaded: true }', () => {
      const wrapper = shallowWithStyles(<ResponsiveImage {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
