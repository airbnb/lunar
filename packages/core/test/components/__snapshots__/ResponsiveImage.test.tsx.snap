// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResponsiveImage /> with a shimmer prop renders the image with state={ imageLoaded: true } 1`] = `
<img
  alt="bar"
  className="image_9yfome-o_O-inline-0_2z4ibo"
  height="auto"
  src="foo"
  width="100%"
/>
`;

exports[`<ResponsiveImage /> with a shimmer prop renders the shimmer component while the image is loading 1`] = `
<Shimmer
  block={true}
/>
`;

exports[`<ResponsiveImage /> without a shimmer prop renders the image element while the image is loading 1`] = `
<img
  alt="bar"
  className="image_9yfome-o_O-inline-0_2z4ibo"
  height="auto"
  src="foo"
  width="100%"
/>
`;

exports[`<ResponsiveImage /> without a shimmer prop renders the image element with state={ imageLoaded: true } 1`] = `
<img
  alt="bar"
  className="image_9yfome-o_O-inline-0_2z4ibo"
  height="auto"
  src="foo"
  width="100%"
/>
`;
